Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs

```diff
@@ -0,0 +1,73 @@
+using Content.Server.Atmos.EntitySystems;
+using Content.Server.Atmos.Piping.Components;
+using Content.Server.Atmos.Piping.Unary.Components;
+using Content.Server.NodeContainer;
+using Content.Server.NodeContainer.EntitySystems;
+using Content.Server.NodeContainer.Nodes;
+using Content.Server.Power.Components;
+using Content.Shared.Atmos;
+using JetBrains.Annotations;
+using Content.Server.Power.EntitySystems;
+using Content.Shared.Chemistry.EntitySystems;
+
+namespace Content.Server.Atmos.Piping.Unary.EntitySystems;
+
+[UsedImplicitly]
+public sealed class GasCondenserSystem : EntitySystem
+{
+    [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+    [Dependency] private readonly PowerReceiverSystem _power = default!;
+    [Dependency] private readonly NodeContainerSystem _nodeContainer = default!;
+    [Dependency] private readonly SolutionContainerSystem _solution = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<GasCondenserComponent, AtmosDeviceUpdateEvent>(OnThermoMachineUpdated);
+    }
+
+    private void OnThermoMachineUpdated(EntityUid uid, GasCondenserComponent component, AtmosDeviceUpdateEvent args)
+    {
+        if (!(_power.IsPowered(uid) && TryComp<ApcPowerReceiverComponent>(uid, out var receiver))
+            || !TryComp<NodeContainerComponent>(uid, out var nodeContainer)
+            || !_nodeContainer.TryGetNode(nodeContainer, component.Inlet, out PipeNode? inlet)
+            || !_solution.TryGetSolution(uid, component.SolutionId, out var solution))
+        {
+            return;
+        }
+
+        if (solution.AvailableVolume == 0 || inlet.Air.TotalMoles == 0)
+            return;
+
+        var molesToConvert = NumberOfMolesToConvert(receiver, inlet.Air, args.dt);
+        var removed = inlet.Air.Remove(molesToConvert);
+
+        for (var i = 0; i < Atmospherics.TotalNumberOfGases; i++)
+        {
+            var moles = removed.Moles[i];
+            if (moles <= 0)
+                continue;
+
+            if (_atmosphereSystem.GetGas(i).Reagent is not {} gasReagent)
+                continue;
+
+            var moleToReagentMultiplier = component.MolesToReagentMultiplier;
+            var amount = moles * moleToReagentMultiplier;
+
+            if (_solution.TryAddReagent(uid, solution, gasReagent, amount, out var remaining))
+                continue;
+
+            // if we have leftover reagent, then convert it back to moles and put it back in the mixture.
+            inlet.Air.AdjustMoles(i, remaining.Float() / moleToReagentMultiplier);
```

Comment:
This technically breaks the conservation of energy, i.e. if you do this too often the inlet will mysteriously cool down for no apparent reason. On the other hand, in my personal opinion I don't think this is worth fixing until it proves to be a problem.

Reply:
Yeah i figured its going to be a relatively small amount of moles and I'd rather preserve the mass at the cost of a small bit of heat leaking out.